flake.nix

```diff
@@ -0,0 +1,70 @@
+/*
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+{
+  description = "XiangShan";
+
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
+    flake-utils.url = "github:numtide/flake-utils";
+  };
+
+  outputs = { self, nixpkgs, flake-utils }@inputs:
+    let
+      overlay = import ./overlay.nix;
+    in
+    flake-utils.lib.eachDefaultSystem
+      (system:
+      let
+        pkgs = import nixpkgs { inherit system; overlays = [ overlay ]; };
+        deps = with pkgs; [
+          git
+          glibcLocales
+          gnumake autoconf automake
+          mill
+          dtc
+          verilator-5016 cmake ninja
+          python3 python3Packages.pip
+          pkgsCross.riscv64-embedded.buildPackages.gcc
+          pkgsCross.riscv64-embedded.buildPackages.gdb
+          openocd
+          circt
+          mold
+          scons
+          gcc
+          sqlite
+          zlib
+
+        ];
+      in
+        {
+          legacyPackages = pkgs;
+          devShell = pkgs.mkShell.override { stdenv = pkgs.clangStdenv; } {
+            buildInputs = deps;
+            RV64_TOOLCHAIN_ROOT = "${pkgs.pkgsCross.riscv64-embedded.buildPackages.gcc}";
+            shellHook = ''
+              # Tells pip to put packages into $PIP_PREFIX instead of the usual locations.
+              # See https://pip.pypa.io/en/stable/user_guide/#environment-variables.
+              export PIP_PREFIX=$(pwd)/venv/pip_packages
+              export PYTHONPATH="$PIP_PREFIX/${pkgs.python3.sitePackages}:$PYTHONPATH"
+              export PATH="$PIP_PREFIX/bin:$PATH"
+              export NOOP_HOME=$(pwd)
+              unset SOURCE_DATE_EPOCH
+              pip3 install importlib-metadata typing-extensions riscof==1.25.2 pexpect psutil xlsxwriter
```

Comment:
Is riscof required? I don't think xiangshan is using riscof now.

For other packages, do we need them?

Reply:
This is copy&pasted from RC, I don't think Xiangshan deps on RISC-V Arch Test.

Reply:
The first three are copied from rocket-chip.
Some Python scripts in `scripts/` (mainly top-down related) depends on `numpy` and `pandas`. Not sure if we want them in nix env.

Reply:
> This is copy&pasted from RC, I don't think Xiangshan deps on RISC-V Arch Test.

riscv-arch-test is running as bare metal on xiangshan. We are able to build them alone, no need for riscof

Reply:
riscof is like third-party RISC-V library, so we need to be careful when importing them.  I think Python libraries are safe but these third-party risc-v projects should not be easily merged into xiangshan.